chatkit/src/main/java/com/pusher/chatkit/UserSubscription.kt

```diff
@@ -81,98 +82,51 @@ class UserSubscription(
+            path = "/cursors/0/users/$userId"
+        ),
+        tokenProvider = tokenProvider,
+        tokenParams = tokenParams
+    ).parseResponseWhenReady<Array<Cursor>>()
+        .mapResult { cursors -> cursors.map { it.roomId to it }.toMap() }
 
     private fun handleEvent(event: SubscriptionEvent) {
-
-        logger.verbose("Handle Event $event")
-
-        val chatEvent = ChatManager.GSON.fromJson<ChatEvent>(event.body, ChatEvent::class.java)
-        when (chatEvent.eventName) {
-            "initial_state" -> {
-                val body = ChatManager.GSON.fromJson<InitialState>(chatEvent.data, InitialState::class.java)
-                handleInitialState(body)
-            }
-            "added_to_room" -> {
-                val body = ChatManager.GSON.fromJson<RoomEvent>(chatEvent.data, RoomEvent::class.java)
-                handleAddedToRoom(body.room)
-            }
-            "removed_from_room" -> {
-                val body = ChatManager.GSON.fromJson<RoomIdEvent>(chatEvent.data, RoomIdEvent::class.java)
-                handleRemovedFromRoom(body.roomId)
-            }
-            "room_updated" -> {
-                val body = ChatManager.GSON.fromJson<RoomEvent>(chatEvent.data, RoomEvent::class.java)
-                handleRoomUpdated(body.room)
-            }
-            "room_deleted" -> {
-                val body = ChatManager.GSON.fromJson<RoomIdEvent>(chatEvent.data, RoomIdEvent::class.java)
-                handleRoomDeleted(body.roomId)
-            }
-            "user_joined" -> {
-                val body = ChatManager.GSON.fromJson<UserChangeEvent>(chatEvent.data, UserChangeEvent::class.java)
-                handleUserJoined(body.userId, body.roomId)
-
-            }
-            "user_left" -> {
-                val body = ChatManager.GSON.fromJson<UserChangeEvent>(chatEvent.data, UserChangeEvent::class.java)
-                handleUserLeft(body.userId, body.roomId)
-            }
-            else -> {
-                throw Error("Invalid event name: ${chatEvent.eventName}")
+        event.body.parseAs<ChatEvent>()
+            .map { (eventName, _, _, data) ->
+                when (eventName) {
+                    "initial_state" -> data.parseAs<InitialState>().map { handleInitialState(it) }
+                    "added_to_room" -> data.parseAs<RoomEvent>().map { handleAddedToRoom(it.room) }
+                    "removed_from_room" -> data.parseAs<RoomIdEvent>().map { handleRemovedFromRoom(it.roomId) }
+                    "room_updated" -> data.parseAs<RoomEvent>().map { handleRoomUpdated(it.room) }
+                    "room_deleted" -> data.parseAs<RoomIdEvent>().map { handleRoomDeleted(it.roomId) }
+                    "user_joined" -> data.parseAs<UserChangeEvent>().map { handleUserJoined(userId, it.roomId) }
+                    "user_left" -> data.parseAs<UserChangeEvent>().map { handleUserLeft(userId, it.roomId) }
+                    else -> throw Error("Invalid event name: $eventName")
```

Comment:
`Unknown event name`

Which begs the question: how will graceful will this SDK handle new events being added?